### PR TITLE
Node 1442 arbitrary asset in exchange transactions fee

### PR DESCRIPTION
**Pull request by DenisSerebryansky**

### CODE REVIEW

File: src/test/scala/com/wavesplatform/transaction/ExchangeTransactionSpecification.scala [hunk 1]
```diff
@@ -16,6 +17,33 @@ import scala.math.pow
...
+      (1: Byte, 1: Byte, 2: Byte),
+      (1: Byte, 2: Byte, 2: Byte),
+      (1: Byte, 3: Byte, 2: Byte),
+      (2: Byte, 1: Byte, 2: Byte),
+      (2: Byte, 2: Byte, 2: Byte),
+      (2: Byte, 3: Byte, 2: Byte),
+      (3: Byte, 1: Byte, 2: Byte),
+      (3: Byte, 2: Byte, 2: Byte),
+      (3: Byte, 3: Byte, 2: Byte)
+    )
```
Comment (vsuharnikov) [inline src/test/scala/com/wavesplatform/transaction/ExchangeTransactionSpecification.scala]: for-comprehension?

File: src/main/scala/com/wavesplatform/state/diffs/ExchangeTransactionDiff.scala [hunk 2]
```diff
@@ -121,4 +147,11 @@ object ExchangeTransactionDiff {
     else if (!sellFeeValid) Left(OrderValidationError(exTrans.sellOrder, s"Insufficient sell fee"))
     else Right(exTrans)
   }
+
+  def wavesPortfolio(amt: Long) = Portfolio(amt, LeaseBalance.empty, Map.empty)
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply src/main/scala/com/wavesplatform/state/diffs/ExchangeTransactionDiff.scala]: Done

File: src/main/scala/com/wavesplatform/transaction/assets/exchange/Order.scala [hunk 3]
```diff
@@ -258,14 +261,31 @@ object Order {
...
     val sig = crypto.sign(sender, unsigned.bodyBytes())
+
     unsigned match {
-      case o @ OrderV2(_, _, _, _, _, _, _, _, _, _) =>
-        o.copy(proofs = Proofs(Seq(ByteStr(sig))))
-      case o @ OrderV1(_, _, _, _, _, _, _, _, _, _) =>
-        o.copy(proofs = Proofs(Seq(ByteStr(sig))))
+      case o: OrderV1 => o.copy(proofs = Proofs(Seq(ByteStr(sig))))
+      case o: OrderV2 => o.copy(proofs = Proofs(Seq(ByteStr(sig))))
+      case o: OrderV3 => o.copy(proofs = Proofs(Seq(ByteStr(sig))))
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply src/main/scala/com/wavesplatform/transaction/assets/exchange/Order.scala]: Done

File: src/test/scala/com/wavesplatform/TransactionGen.scala [hunk 4]
```diff
@@ -633,23 +638,38 @@ trait TransactionGenBase extends ScriptGen with NTPTime { _: Suite =>
                            amountAssetId: Option[ByteStr],
                            priceAssetId: Option[ByteStr],
                            fixedMatcherFee: Option[Long] = None,
-                           orderVersions: Set[Byte] = Set(1, 2)): Gen[ExchangeTransactionV2] = {
-    def mkBuyOrder(version: Byte): OrderConstructor  = if (version == 1) OrderV1.buy else OrderV2.buy
-    def mkSellOrder(version: Byte): OrderConstructor = if (version == 1) OrderV1.sell else OrderV2.sell
+                           orderVersions: Set[Byte] = Set(1, 2, 3),
+                           buyMatcherFeeAssetId: Option[ByteStr] = None,
+                           sellMatcherFeeAssetId: Option[ByteStr] = None,
+                           fixedMatcher: Option[PrivateKeyAccount] = None): Gen[ExchangeTransactionV2] = {
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply src/test/scala/com/wavesplatform/TransactionGen.scala]: Done

File: src/test/scala/com/wavesplatform/transaction/ExchangeTransactionSpecification.scala (hunk 1 above)
Comment (DenisSerebryansky) [reply src/test/scala/com/wavesplatform/transaction/ExchangeTransactionSpecification.scala]: Done

File: src/main/scala/com/wavesplatform/database/LevelDBWriter.scala [hunk 5]
```diff
@@ -375,9 +380,9 @@ class LevelDBWriter(writableDB: DB,
...
       val minVotes = fs.blocksForFeatureActivation(height)
-      val newlyApprovedFeatures = featureVotes(height).collect {
-        case (featureId, voteCount) if voteCount + (if (block.featureVotes(featureId)) 1 else 0) >= minVotes => featureId -> height
-      }
+      val newlyApprovedFeatures = featureVotes(height)
+        .filterNot { case (featureId, _) => fs.preActivatedFeatures.contains(featureId) }
+        .collect { case (featureId, voteCount) if voteCount + (if (block.featureVotes(featureId)) 1 else 0) >= minVotes => featureId -> height }
 
       if (newlyApprovedFeatures.nonEmpty) {
         approvedFeaturesCache = newlyApprovedFeatures ++ rw.get(Keys.approvedFeatures)
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply src/main/scala/com/wavesplatform/database/LevelDBWriter.scala]: Done

File: it/src/test/scala/com/wavesplatform/it/sync/transactions/SignAndBroadcastApiSuite.scala [hunk 6]
```diff
@@ -343,12 +344,22 @@ class SignAndBroadcastApiSuite extends BaseTransactionSuite with NTPTime {
...
+           (2: Byte, 1: Byte, 2: Byte, None, None),
+           (2: Byte, 2: Byte, 2: Byte, None, None),
+           (3: Byte, 1: Byte, 2: Byte, None, None),
+           (3: Byte, 1: Byte, 2: Byte, assetId, None),
+           (3: Byte, 2: Byte, 2: Byte, None, None),
+           (3: Byte, 2: Byte, 2: Byte, assetId, None),
+           (1: Byte, 3: Byte, 2: Byte, None, None),
+           (2: Byte, 3: Byte, 2: Byte, None, None),
+           (3: Byte, 3: Byte, 2: Byte, None, None),
+           (3: Byte, 3: Byte, 2: Byte, assetId, None),
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply it/src/test/scala/com/wavesplatform/it/sync/transactions/SignAndBroadcastApiSuite.scala]: Done

File: it/src/test/scala/com/wavesplatform/it/sync/transactions/ExchangeTransactionSuite.scala [hunk 7]
```diff
@@ -95,4 +104,96 @@ class ExchangeTransactionSuite extends BaseTransactionSuite with NTPTime {
...
+    nodes.waitForHeightAriseAndTxPresent(assetId.base58)
+
+    for ((o1ver, o2ver, matcherFeeOrder1, matcherFeeOrder2) <- Seq(
+           (1: Byte, 3: Byte, None, Some(assetId)),
+           (2: Byte, 3: Byte, None, Some(assetId)),
+           (1: Byte, 3: Byte, None, None),
+           (3: Byte, 1: Byte, Some(assetId), None),
+           (2: Byte, 3: Byte, None, None),
+           (3: Byte, 2: Byte, Some(assetId), None),
+         )) {
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply it/src/test/scala/com/wavesplatform/it/sync/transactions/ExchangeTransactionSuite.scala]: It seems to be adhoc solution, dont think that using `for-comprehension` would clarify the code

File: it/src/test/scala/com/wavesplatform/it/sync/transactions/ExchangeTransactionSuite.scala [hunk 8]
```diff
@@ -3,51 +3,60 @@ package com.wavesplatform.it.sync.transactions
...
     for ((o1ver, o2ver, tver) <- Seq(
            (1: Byte, 1: Byte, 1: Byte),
            (1: Byte, 1: Byte, 2: Byte),
            (1: Byte, 2: Byte, 2: Byte),
            (2: Byte, 1: Byte, 2: Byte),
-           (2: Byte, 2: Byte, 2: Byte)
+           (2: Byte, 2: Byte, 2: Byte),
+           (3: Byte, 1: Byte, 2: Byte),
+           (2: Byte, 3: Byte, 2: Byte),
+           (3: Byte, 3: Byte, 2: Byte)
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply it/src/test/scala/com/wavesplatform/it/sync/transactions/ExchangeTransactionSuite.scala]: Done

File: it/src/test/scala/com/wavesplatform/it/sync/smartcontract/ExchangeWithContractsSuite.scala [hunk 9]
```diff
@@ -80,10 +79,15 @@ class ExchangeWithContractsSuite extends BaseTransactionSuite with CancelAfterFa
         (contr2, acc1),
         (mcontr, acc2),
       )
+      for ((o1ver, o2ver) <- Seq(
+             (2: Byte, 2: Byte),
+             (2: Byte, 3: Byte),
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply it/src/test/scala/com/wavesplatform/it/sync/smartcontract/ExchangeWithContractsSuite.scala]: It seems to be adhoc, clear enough imho

File: it/src/test/scala/com/wavesplatform/it/sync/smartcontract/ExchangeWithContractsSuite.scala [hunk 10]
```diff
@@ -80,10 +79,15 @@ class ExchangeWithContractsSuite extends BaseTransactionSuite with CancelAfterFa
...
+           )) {
 
-      sender.signedBroadcast(exchangeTx(pair, smartMatcherFee, orderFee, ntpTime, acc1, acc0, acc2), waitForTx = true)
+        sender.signedBroadcast(exchangeTx(pair, smartMatcherFee, orderFee, ntpTime, o1ver, o2ver, acc1, acc0, acc2), waitForTx = true)
 
-      //TODO : add assert balances
+        //TODO : add assert balances
```
[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply it/src/test/scala/com/wavesplatform/it/sync/smartcontract/ExchangeWithContractsSuite.scala]: No idea

File: it/src/test/scala/com/wavesplatform/it/sync/smartcontract/ExchangeWithContractsSuite.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (vsuharnikov) [reply it/src/test/scala/com/wavesplatform/it/sync/smartcontract/ExchangeWithContractsSuite.scala]: I meant, why there is (2, 2), but not (3, 3) ?

[the remark this answers is not in the document]
Comment (DenisSerebryansky) [reply it/src/test/scala/com/wavesplatform/it/sync/smartcontract/ExchangeWithContractsSuite.scala]: Actually I don't know, it's @monroid commit

**Pull request merged**